contrib/mesos/pkg/scheduler/service/util.go

```diff
@@ -86,3 +94,131 @@ func hashExecutorInfo(info *mesos.ExecutorInfo) uint64 {
+	go func() {
+		defer close(out)
+		for entry := range in {
+			if entry.Dir {
+				continue
+			}
+			if entry.Err != nil {
+				log.Errorf("error fetching static pod file %q: %v", entry.Path, entry.Err)
+				continue
+			}
+			keep := false
+			defaultFn := func(pod *api.Pod) (err error) {
+				keep, err = filter(pod)
+				return
+			}
+			parsed, pod, err := tryDecodeSinglePod(entry.Data, defaultFn)
+			if !parsed {
+				if err != nil {
+					entry.Err = err
+					log.Errorf("error parsing static pod file %q: %v", entry.Path, err)
+				}
+				continue
+			}
+			if err != nil {
+				entry.Err = err
+				log.Errorf("error validating static pod file %q: %v", entry.Path, err)
+				continue
+			}
+			if keep {
+				// marshal back into JSON form to pick up any changes made to
+				// the pod specification
+				updated, err := api.Scheme.EncodeToVersion(pod, "v1")
+				if err != nil {
+					entry.Err = err
+					log.Errorf("error serializing pod json for %v/%v: %v", pod.Namespace, pod.Name, err)
+					continue
+				}
+				entry.Data = updated
+				out <- entry
+			}
+		}
+	}()
+	return out
+}
+
+func makeStaticPodValidator(limitCPU resource.CPUShares, limitMem resource.MegaBytes, accumCPU, accumMem *float64, minimalResources bool) filterFunc {
+	return filterFunc(func(pod *api.Pod) (bool, error) {
+		// TODO(sttts): allow unlimited static pods as well and patch in the default resource limits
+		unlimitedCPU := resource.LimitPodCPU(pod, limitCPU)
+		if unlimitedCPU {
+			return false, fmt.Errorf("found static pod without limit on cpu resources")
+		}
+
+		unlimitedMem := resource.LimitPodMem(pod, limitMem)
+		if unlimitedMem {
+			return false, fmt.Errorf("found static pod without limit on memory resources")
+		}
+
+		cpu := resource.PodCPULimit(pod)
+		mem := resource.PodMemLimit(pod)
+		if minimalResources {
+			// see SchedulerServer.AccountForPodResources
+			cpu = podtask.MinimalCpus
+			mem = podtask.MinimalMem
+		}
+
+		log.V(2).Infof("reserving %.2f cpu shares and %.2f MB of memory to static pod %s/%s", cpu, mem, pod.Namespace, pod.Name)
+
+		*accumCPU += float64(cpu)
+		*accumMem += float64(mem)
+		return true, nil
+	})
+}
+
+func makeStaticPodProcurement(entries []*archive.Entry) podtask.Procurement {
+	if len(entries) == 0 {
+		return nil
+	}
+
+	// entries is a pre-validated list of static pod specs that serve as templates for the static pod builder func
+	return podtask.StaticPodProcurement(func(assignedSlave string) ([]byte, error) {
+		applyBindingsTo := filterFunc(func(pod *api.Pod) (bool, error) {
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string)
+			}
+			// add binding annotation so that static pods aren't picked up by the main scheduling loop.
+			// static pods are always completely managed by the kubelet.
+			pod.Annotations[meta.BindingHostKey] = assignedSlave
+			return true, nil
+		})
+		return archive.Create(applyBindingsTo.podEntries(archive.Playback(entries)))
```

Comment:
If we did the pod list thing I'd be tempted to capture the file name in an
annotation so that when we create the static pods dir on the executor, the
layout makes sense to someone trying to debug why their static pod may not
be working as expected (not our fault, hopefully!). But then what's the
difference between a compressed list of pod entries that know their file
names and a zip archive? The complexity feels almost the same to me. The
zip archive adds complexity in that it supports a multi-layer filesystem
structure (that we don't really need) --- but that's somewhat trivial in
the code; and we also add an extra deserialize/serialize step since the
pods in the zip container are stored in byte form (we only pay this price
once per slave).

I guess I'm questioning whether it's worth duplicating some of the
functionality that zip gives us by managing a pod-list.

Will think on this more.


On Fri, Sep 18, 2015 at 10:26 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/scheduler/service/util.go
> <https://github.com/kubernetes/kubernetes/pull/13943#discussion_r39859756>
> :
>
> > +	if len(entries) == 0 {
> > +		return nil
> > +	}
> > +
> > +	// entries is a pre-validated list of static pod specs that serve as templates for the static pod builder func
> > +	return podtask.StaticPodProcurement(func(assignedSlave string) ([]byte, error) {
> > +		applyBindingsTo := filterFunc(func(pod *api.Pod) (bool, error) {
> > +			if pod.Annotations == nil {
> > +				pod.Annotations = make(map[string]string)
> > +			}
> > +			// add binding annotation so that static pods aren't picked up by the main scheduling loop.
> > +			// static pods are always completely managed by the kubelet.
> > +			pod.Annotations[meta.BindingHostKey] = assignedSlave
> > +			return true, nil
> > +		})
> > +		return archive.Create(applyBindingsTo.podEntries(archive.Playback(entries)))
>
> I don't say to use another file source. We can still write the json
> snippets out into flat files. But for the transport and the patching we
> don't need a zip. The file source also does not support directories. So
> there is no need for this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13943/files#r39859756>.
>